In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


To-Do-1

In [ ]:
import pandas as pd

#1 Read and Observe the Dataset.
file = "/content/drive/MyDrive/Datasets-20241203T015332Z-001/Datasets/student.csv"
df = pd.read_csv(file)

#2 Print top(5) and bottom(5) of the dataset
print("Top 5 of dataset: \n", df.head(5))
print("\n")
print("Bottom 5 of dataset: \n", df.tail(5))
print("\n")

#3 Print the Information of Datasets.
print("Information of dataset: \n", df.info)
print("\n")

#4 Gather the Descriptive info about the Dataset.
print("Description of dataset: \n", df.describe)
print("\n")

#5 Split your data into Feature (X) and Label (Y).
X = df[['Math', 'Reading']]
Y = df['Writing']
print("\nData has been split into features and labels.")
print("Features (X):")
print(X.head())
print("\nLabel (Y):")
print(Y.head())

Top 5 of dataset: 
    Math  Reading  Writing
0    48       68       63
1    62       81       72
2    79       80       78
3    76       83       79
4    59       64       62


Bottom 5 of dataset: 
      Math  Reading  Writing
995    72       74       70
996    73       86       90
997    89       87       94
998    83       82       78
999    66       66       72


Information of dataset: 
 <bound method DataFrame.info of      Math  Reading  Writing
0      48       68       63
1      62       81       72
2      79       80       78
3      76       83       79
4      59       64       62
..    ...      ...      ...
995    72       74       70
996    73       86       90
997    89       87       94
998    83       82       78
999    66       66       72

[1000 rows x 3 columns]>


Description of dataset: 
 <bound method NDFrame.describe of      Math  Reading  Writing
0      48       68       63
1      62       81       72
2      79       80       78
3      76       83       79
4      

To-Do-2

In [ ]:
import pandas as pd
import numpy as np

file = "/content/drive/MyDrive/Datasets-20241203T015332Z-001/Datasets/student.csv"
df = pd.read_csv(file)

X = df[['Math', 'Reading']].values.T
W = np.random.randn(X.shape[0], 1)
Y = np.dot(W.T, X)

print("Feature matrix X (d x n):")
print(X)

print("\nWeights W (d x 1):")
print(W)

print("\nOutput Matrix Y :")
print(Y)

Feature matrix X (d x n):
[[48 62 79 ... 89 83 66]
 [68 81 80 ... 87 82 66]]

Weights W (d x 1):
[[ 2.09529297]
 [-0.42443215]]

Output Matrix Y :
[[ 71.71267658  95.52916024 131.57357284 124.0143975   96.45862781
  108.92291458 111.01820755  76.01073355  94.70716371 147.0894879
  100.62234598 113.5110649   98.66139181 136.61302306  98.10262087
   83.99434103  95.58289576  73.91544058 110.56690764 125.31456169
  113.67227145  91.41917758 109.82551439 109.93298542 105.97875732
  104.28102874 101.89564242  85.50399531 138.70831603  99.29531403
  116.08452553 106.37632171 102.69077119 129.95644753 110.41115308
   72.58840863  82.26974469 124.4656974  113.16723603  69.32729026
  116.03079002 130.75157631 109.8792499   86.06276624  98.95148516
  114.01610032  84.36503766 102.69077119 146.13315258  90.11901339
  167.08608226 115.18192573  95.52916024  72.64214414  77.25716223
   90.25335218 106.56439601  81.44774816 132.0517405  112.29150398
  102.37381008 132.766266   167.93494655 115.63322

To-Do-3

In [ ]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split

file = "/content/drive/MyDrive/Datasets-20241203T015332Z-001/Datasets/student.csv"
df = pd.read_csv(file)

train_ratio = 0.8
train_size = int(len(df) * train_ratio)

train_data = df.iloc[:train_size]
test_data = df.iloc[train_size:]

X_train = train_data.drop('Math', axis=1)
y_train = train_data['Math']

X_test = test_data.drop('Math', axis=1)
y_test = test_data['Math']

print(f"Training data size (X_train): {len(X_train)} rows")
print(f"Test data size (X_test): {len(X_test)} rows")

Training data size (X_train): 800 rows
Test data size (X_test): 200 rows


To-Do

In [7]:
import numpy as np
import pandas as pd

def gradient_descent(X, y, W, learning_rate, num_iterations):
    m = len(y)
    for i in range(num_iterations):
        predictions = X.dot(W)
        errors = predictions - y
        gradient = (1/m) * X.T.dot(errors)
        W -= learning_rate * gradient
        if (i+1) % 100 == 0:
            loss = np.mean(errors**2)
            print(f"Iteration {i+1}: Loss = {loss}")
    return W

def rmse(y_true, y_pred):
    return np.sqrt(np.mean((y_true - y_pred) ** 2))

def r2_score_custom(y_true, y_pred):
    total_variance = np.sum((y_true - np.mean(y_true)) ** 2)
    unexplained_variance = np.sum((y_true - y_pred) ** 2)
    return 1 - (unexplained_variance / total_variance)

def train_and_evaluate_model(data, target_column, learning_rate=0.01, num_iterations=1000, train_ratio=0.8):
    X = data.drop(target_column, axis=1).values
    y = data[target_column].values.reshape(-1, 1)
    X = (X - X.mean(axis=0)) / X.std(axis=0)
    X = np.c_[np.ones((X.shape[0], 1)), X]
    train_size = int(len(data) * train_ratio)
    X_train, X_test = X[:train_size], X[train_size:]
    y_train, y_test = y[:train_size], y[train_size:]
    W = np.zeros((X_train.shape[1], 1))
    print("Starting Gradient Descent...")
    W = gradient_descent(X_train, y_train, W, learning_rate, num_iterations)
    print("Gradient Descent Completed.\n")
    y_pred = X_test.dot(W)
    rmse_value = rmse(y_test, y_pred)
    r2_value = r2_score_custom(y_test, y_pred)
    print(f"Model Evaluation Results:")
    print(f"RMSE (Root Mean Squared Error): {rmse_value:.4f}")
    print(f"R² (Coefficient of Determination): {r2_value:.4f}\n")
    return W

df = pd.read_csv("/content/drive/MyDrive/Datasets-20241203T015332Z-001/Datasets/student.csv")
W = train_and_evaluate_model(df, target_column='Math', learning_rate=0.01, num_iterations=1000)


Starting Gradient Descent...
Iteration 100: Loss = 699.2962963635672
Iteration 200: Loss = 162.49871532444774
Iteration 300: Loss = 90.90732975530241
Iteration 400: Loss = 81.3166534173156
Iteration 500: Loss = 80.02741601616349
Iteration 600: Loss = 79.85066715182437
Iteration 700: Loss = 79.82332612061171
Iteration 800: Loss = 79.81633321119472
Iteration 900: Loss = 79.81236775342418
Iteration 1000: Loss = 79.80908195225282
Gradient Descent Completed.

Model Evaluation Results:
RMSE (Root Mean Squared Error): 8.0630
R² (Coefficient of Determination): 0.6688



With an R² of 0.67, the model demonstrates moderate predictive power, indicating a reasonable fit to the data, though it is not flawless.

The RMSE of 8.06 indicates a moderate level of error between predictions and actual values. However, without understanding the scale of the target variable, it is challenging to determine if this error is considered large or acceptable.

In [9]:
# Experiment with different value of learning rate
learning_rates = [0.001, 0.01, 0.1, 0.5, 1]

for lr in learning_rates:
    print(f"Evaluating model with learning rate: {lr}")
    W = train_and_evaluate_model(df, target_column='Math', learning_rate=lr, num_iterations=1000)
    print("-" * 50)

Evaluating model with learning rate: 0.001
Starting Gradient Descent...
Iteration 100: Loss = 3889.0670099762856
Iteration 200: Loss = 3181.745091095462
Iteration 300: Loss = 2608.1846780395454
Iteration 400: Loss = 2142.3042431056147
Iteration 500: Loss = 1763.3582670188446
Iteration 600: Loss = 1454.7681111661116
Iteration 700: Loss = 1203.2318619387745
Iteration 800: Loss = 998.0403036668321
Iteration 900: Loss = 830.546782573105
Iteration 1000: Loss = 693.7531351203352
Gradient Descent Completed.

Model Evaluation Results:
RMSE (Root Mean Squared Error): 26.4877
R² (Coefficient of Determination): -2.5742

--------------------------------------------------
Evaluating model with learning rate: 0.01
Starting Gradient Descent...
Iteration 100: Loss = 699.2962963635672
Iteration 200: Loss = 162.49871532444774
Iteration 300: Loss = 90.90732975530241
Iteration 400: Loss = 81.3166534173156
Iteration 500: Loss = 80.02741601616349
Iteration 600: Loss = 79.85066715182437
Iteration 700: Loss =

Lower Learning Rate (e.g., 0.001): Training will take longer, but the process may be more stable, avoiding overshooting. If the model underperforms, it could be due to insufficient learning from slow updates.

Moderate Learning Rate (e.g., 0.01): Strikes a balance between training speed and convergence. This is typically a solid starting point for many models.

Higher Learning Rate (e.g., 0.1, 0.5, 1): Training may be faster, but there's a greater risk of overshooting the optimal weights. If the learning rate is too high, the model might not converge properly, or it could result in significant errors (high RMSE).






